sbncode/LArRecoProducer/CRUMBS_module.cc

```diff
@@ -0,0 +1,681 @@
+      fSliceTree->Branch("runID",&runID);
+      fSliceTree->Branch("slicePDG",&slicePDG);
+      fSliceTree->Branch("matchedIndex",&matchedIndex);
+      fSliceTree->Branch("matchedType",&matchedType);
+      fSliceTree->Branch("matchedPurity",&matchedPurity);
+      fSliceTree->Branch("matchedCompleteness",&matchedCompleteness);
+    }
+
+  void CRUMBS::ClearMaps() 
+  {
+    fTrackIDToGenMap.clear();
+    fGenTypeMap.clear();
+  }
+
+  void CRUMBS::ResetVars()
+  {
+    tpc_NuScore = -999999.; tpc_CRFracHitsInLongestTrack = -999999.; tpc_CRLongestTrackDeflection = -999999.; tpc_CRLongestTrackDirY = -999999.; tpc_CRNHitsMax = -999999.;
+    tpc_NuEigenRatioInSphere = -999999.; tpc_NuNFinalStatePfos = -999999.; tpc_NuNHitsTotal = -999999.; tpc_NuNSpacePointsInSphere = -999999.; tpc_NuVertexY = -999999.;
+    tpc_NuWeightedDirZ = -999999.; tpc_StoppingChi2CosmicRatio = -4.;
+
+    pds_FMTotalScore = -999999.; pds_FMPE = -999999.; pds_FMTime = -500.;
+
+    crt_TrackScore = -4.; crt_HitScore = -4.; crt_TrackTime = -3000; crt_HitTime = -3000;
+
+    slicePDG = 999999; matchedIndex = 999999;
+    matchedType = "";
+    matchedPurity = -999999.; matchedCompleteness = -999999.;
+  }
+
+  void CRUMBS::SetupMaps(art::Event const& e)
+  {
+
+    unsigned nNu(0), nCos(0);
+
+    if(fProcessNeutrinos)
+      {
+	art::Handle<std::vector<simb::MCTruth> > handleMCTruthNu;
+	e.getByLabel(fGeneratorModuleLabel, handleMCTruthNu);
+	art::FindManyP<simb::MCParticle> truthNuMCPAssn(handleMCTruthNu,e,fMCParticleModuleLabel);
+
+	for (unsigned int i = 0; i < handleMCTruthNu->size(); ++i){
+	  const art::Ptr<simb::MCTruth> mcTruth(handleMCTruthNu, i);
+	  const simb::MCParticle nu = mcTruth->GetNeutrino().Nu();
+
+	  if(!fTpcGeo.InVolume(nu))
+	    fGenTypeMap[i] = "DirtNu";
+	  else
+	    fGenTypeMap[i] = "Nu";
+    
+	  const std::vector<art::Ptr<simb::MCParticle> > particles = truthNuMCPAssn.at(mcTruth.key());
+    
+	  for (auto const& particle : particles)
+	    {
+	      fTrackIDToGenMap[particle->TrackId()] = i;
+	    }
+	  ++nNu;
+	}
+      }
+
+    if(fProcessCosmics)
+      {
+	art::Handle<std::vector<simb::MCTruth> > handleMCTruthCosmic;
+	e.getByLabel(fCosmicModuleLabel, handleMCTruthCosmic);
+
+	art::FindManyP<simb::MCParticle> truthCosmicMCPAssn(handleMCTruthCosmic,e,fMCParticleModuleLabel);
+
+	for (unsigned int i = 0; i < handleMCTruthCosmic->size(); ++i){
+	  const art::Ptr<simb::MCTruth> mcTruth(handleMCTruthCosmic, i);
+
+	  fGenTypeMap[i + nNu] = "Cosmic";
+    
+	  const std::vector<art::Ptr<simb::MCParticle> > particles = truthCosmicMCPAssn.at(mcTruth.key());
+    
+	  for (auto const& particle : particles)
+	    {
+	      fTrackIDToGenMap[particle->TrackId()] = i + nNu;
+	    }
+	  ++nCos;
+	}
+      }
+  
+    eventID = e.event();
+    subRunID = e.subRun();
+    runID = e.run();
+  }
+
+  void CRUMBS::produce(art::Event& e)
+  {
+    if(fTrainingMode)
+      {
+	this->ClearMaps();
+	this->SetupMaps(e);
+      }
```

Comment:
These member variables, meanwhile, are not necessary. You could have one `GetMaps()`-like function and keep the maps as local variables in this function, and pass them into users. Makes the intent clearer, and prevents any possibility of accidental contamination from one event to the next.

Reply:
Okay will do so!